src/main/java/net/minecraftforge/common/smelting/AbstractSmeltingRecipe.java

```diff
@@ -0,0 +1,41 @@
+package net.minecraftforge.common.smelting;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import net.minecraft.item.ItemStack;
+
+/**
+ * Base class for {@code SmeltingRecipe} implementations.
+ */
+public abstract class AbstractSmeltingRecipe implements SmeltingRecipe
+{
+
+    protected final int duration;
+
+    protected AbstractSmeltingRecipe(int duration)
+    {
+        checkArgument(duration > 0);
+        this.duration = duration;
+    }
+
+    @Override
+    public final ItemStack getOutput(ItemStack input)
+    {
+        return matches(input) ? ItemStack.copyItemStack(getOutput0(input)) : null;
+    }
+
+    /**
+     * Get the raw output for this recipe based on the given input. The resulting stack should not be
+     * a defensive copy. This method is only called, if {@code matches} returns true.
+     *
+     * @param input the input stack
+     * @return the raw output stack
+     */
+    protected abstract ItemStack getOutput0(ItemStack input);
```

Comment:
Maybe a name like 'getRawOutput' would fit better here. Generally i dislike method names with numbers suffixed in public api. Adding to that, i can't really come up with a case where the recipe input does not match, but still wants to generate output. Maybe drop this method and make the normal 'getOutput' method do that?